### PR TITLE
Move File and IO::FileDescriptor's platform-specific parts to Crystal::System

### PR DESCRIPTION
**Pull request by RX14**
Hopefully the third time lucky we'll manage to merge an IO portability refactor.

I've copied the commit messages out below so people actually read them. I also suggest you review each commit in order, instead of reviewing the diff as a whole. it makes the entire thing a lot more readable and understandable.

## Make IO::FileDescriptor fcntl private
fcntl() is an extremely low-level platform-specific detail which should only be used internally to IO::FileDescriptor. In this commit, it is removed from the public API and made private.

The only usage of this API was added only recently, in Crystal::Main. It is replaced with a method which calls LibC.fcntl directly, since it is too early in init to create a new IO::FileDescriptor instance directly and ask it if it's blocking.

## Separate platform-specific parts of IO::FileDescriptor
In order to add windows support to Crystal, IO::FileDescriptor cannot depend on posix APIs. Here, we refactor out the platform-specific parts of IO::FileDescriptor into a mixin module in Crystal::System. Some platform-specific methods are refactored out of methods which contain both platform-specific and cross-platform behaviour. These methods are prefixed with `system_` and are private.

## Make Tempfile inherit File
Making Tempfile inherit from File allows the instance to be passed to methods which expect File, and makes Tempfile much more powerful. This requires adding a new private constructor to File which sets @path and calls super.

## Separate platform-specific parts of File
Similar to previous patch: Separate platform-specific parts of IO::FileDescriptor.

### CODE REVIEW

**Comment by straight-shoota** [comment]
In the second commit, the extracted platform-specific methods are prefixed `platform_`. However, in the commit message it says to be `system_`.

I'd prefer `system_` because it matches the namespace and is more descriptive.

File: src/crystal/system/unix/file.cr [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+        case mode[1]
+        when '+'
+          m = LibC::O_RDWR
+        when 'b'
+          # Nothing
+        else
+          raise "Invalid access mode #{mode}"
+        end
+      else
+        raise "Invalid access mode #{mode}"
```
Comment (Papierkorb) [inline src/crystal/system/unix/file.cr]: Doesn't support a mode like `rb+`. Also, `#{mode.inspect}` would make it clearer in the output what comes from the user and what's the error message itself

File: src/crystal/system/unix/file.cr [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+      oflag = m | o
+    end
+
+    def mktemp(name, extension)
+      tmpdir = tempdir + ::File::SEPARATOR
+      path = "#{tmpdir}#{name}.XXXXXX#{extension}"
+
+      if extension
+        fd = LibC.mkstemps(path, extension.bytesize)
```
Comment (Papierkorb) [inline src/crystal/system/unix/file.cr]: 1. Missing `String#check_no_null_byte`
2. mkstemps(3) reads: `Since it will be modified, template must not be a string constant, but should be declared as a character array.` That's not obvious from the function itself (At least to a non-C user), there should be atleast a comment noting that the string will be modified (Which isn't "possible" from Crystal), or it could be copied into a temporary `Bytes`.

File: src/crystal/system/unix/file.cr [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    end
+
+    def tempdir
+      tmpdir = ENV["TMPDIR"]? || "/tmp"
+      tmpdir.rchop(::File::SEPARATOR)
+    end
+
+    def stat(path)
+      if LibC.stat(path.check_no_null_byte, out stat) != 0
+        raise Errno.new("Unable to get stat for '#{path}'")
```
Comment (Papierkorb) [inline src/crystal/system/unix/file.cr]: `path.inspect` would give proper visual "escaping"

File: src/crystal/system/unix/file.cr [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+    def delete(path)
+      err = LibC.unlink(path.check_no_null_byte)
+      if err == -1
+        raise Errno.new("Error deleting file '#{path}'")
+      end
+    end
+
+    def real_path(path)
+      real_path_ptr = LibC.realpath(path, nil)
+      raise Errno.new("Error resolving real path of #{path}") unless real_path_ptr
```
Comment (Papierkorb) [inline src/crystal/system/unix/file.cr]: Same `path.inspect`, this and following error messages omit the single-quote. (Consistency)

**Comment by Papierkorb** [comment]
I think there needs to be a proper support for platform-specific functionality in the stdlib. IMHO, throwing stuff out like `#fcntl` (The easy access to it) can be harmful: The wrapper methods are usually much easier and nicer to use; The `LibC` pendants are harder to get right.

Truth is, programs tend to need platform-specific stuff that's not yet nicely wrapped.  I doubt that you'll be able to offer everything that the control functions support in an agnostic manner.  And many programs are fine if they don't work on some random platform, as long they work where they're intended to be used.  Diminishing all API to the most common denominator will only hurt on the long run.

Outside the scope of this PR, I'd actually like to see a documentation flag to note platform information right in the in-source docs.  The docs generator could then mark such a section visually, or add a "Tag badge" to it.

File: src/crystal/system/unix/file.cr (hunk 1 above)
Comment (RX14) [reply src/crystal/system/unix/file.cr]: This is directly copied from the old `file.cr` if it has issues, it's out of the scope of this PR.

File: src/crystal/system/unix/file.cr (hunk 2 above)
Comment (RX14) [reply src/crystal/system/unix/file.cr]: ditto

File: src/crystal/system/unix/file.cr (hunk 3 above)
Comment (RX14) [reply src/crystal/system/unix/file.cr]: ditto

File: src/crystal/system/unix/file.cr (hunk 4 above)
Comment (RX14) [reply src/crystal/system/unix/file.cr]: ditto

**Comment by RX14** [comment]
re: removing `fcntl`, I actually think you're right. I'll probably revert that commit.

**Comment by RX14** [comment]
I appreciate these comments, but comments (in support of or against) about the approach of using a mixin module from `Crystal::System` would be most helpful to tell whether this PR is going to get merged quickly.

File: src/crystal/system/unix/file.cr (hunk 1 above)
Comment (Papierkorb) [reply src/crystal/system/unix/file.cr]: Fair enough

File: src/file.cr [hunk 5]
```diff
@@ -654,6 +548,44 @@ class File < IO::FileDescriptor
     io << ">"
     io
   end
+
+  # TODO: use fcntl/lockf instead of flock (which doesn't lock over NFS)
+  # TODO: always use non-blocking locks, yield fiber until resource becomes available
+
+  def flock_shared(blocking = true)
```
[the remark this answers is not in the document]
Comment (RX14) [reply src/file.cr]: I'd tentatively say that's easily done in a latter PR - and so it should be.

[the remark this answers is not in the document]
Comment (ysbaddaden) [reply src/file.cr]: So be it. Let's stick to extracting posix-specific calls in this PR.

**Comment by RX14** [comment]
@ysbaddaden if you look at `File`, there's a `Crystal::System::File` and a `Crystal::System::FileMixin`. The `Mixin` naming comes from there.

File: src/file.cr (hunk 5 above)
[the remark this answers is not in the document]
Comment (RX14) [reply src/file.cr]: I'd recommend making an issue to track that though, we seem to forget a lot of these refactors once the PR gets merged.

**Comment by ysbaddaden** [comment]
Hum, I prefer to use `def self.` instead of `extend self` when there is no actual benefit to using the module methods as both includeable and class methods (for example `Random::Secure` does benefit). This would avoid having `System::File` and `System::FileMixin` and but just `System::File`, with some class methods, and some instance methods to be included in `::File`.

Fairly minor, but it looks better to my eyes.

**Comment by RX14** [comment]
@ysbaddaden sounds good to me. I'll change it.

**Comment by RX14** [comment]
I'm still undecided about naming. I think a `system_` prefix is better than `_impl`, because there's a difference between just a class-private implementation detail and a platform-specific method. `system` corresponds to the namespace, so it'd a good name.

**Comment by ysbaddaden** [comment]
To me it boils down to "implementation detail", but that's pedantic nitpicking, no need to spend hours on such details.

**Comment by RX14** [comment]
Updated this PR. Changed the naming to `system_`, and to remove `Mixin` naming. Also re-added `fcntl` and moved the `IO::System`-specific parts to the mixin too.

**Comment by RX14** [comment]
Also, if this gets merged, please make sure to rebase, not just squash. I think it'll make for a better commit history.

File: src/io/file_descriptor.cr [hunk 6]
```diff
@@ -177,63 +145,14 @@ class IO::FileDescriptor < IO
...
-  private def add_write_event(timeout = @write_timeout)
-    event = @write_event ||= Scheduler.create_fd_write_event(self)
-    event.add timeout
-    nil
-  end
-
   private def unbuffered_rewind
-    seek(0, IO::Seek::Set)
-    self
+    self.pos = 0
```
Comment (ysbaddaden) [inline src/io/file_descriptor.cr]: Why the change?

Comment (straight-shoota) [reply src/io/file_descriptor.cr]: `pos=(value)` calls `seek(value)` and `Seek::Set` is default. So `seek(0)` would be the most direct call here.

File: src/crystal/system/unix/file_descriptor.cr [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+  private def system_seek(offset, whence : IO::Seek) : Nil
+    seek_value = LibC.lseek(@fd, offset, whence)
+
+    if seek_value == -1
+      raise Errno.new "Unable to seek"
+    end
+  end
+
+  private def system_pos
```
Comment (ysbaddaden) [inline src/crystal/system/unix/file_descriptor.cr]: I'd prefer `system_tell` over `system_pos`.

I wish we'd drop `IO#pos` and `IO#pos=` and kept `IO#seek` and `IO#tell` instead (less duplication, more explicit naming).

File: src/crystal/system/unix/file_descriptor.cr [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    arg
+  end
+
+  def self.fcntl(fd, cmd, arg = 0)
+    r = LibC.fcntl(fd, cmd, arg)
+    raise Errno.new("fcntl() failed") if r == -1
+    r
+  end
+
+  def fcntl(cmd, arg = 0)
```
Comment (ysbaddaden) [inline src/crystal/system/unix/file_descriptor.cr]: AFAIK: `IO#fcntl method is supposed to be public, but it won't be documented?

File: src/io/file_descriptor.cr (hunk 6 above)
Comment (RX14) [reply src/io/file_descriptor.cr]: No reason for the change, it's trivial to see it's the same but I can revert it if you want.

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: We just need 1 call: `seek` which always returns the current position. I suggested that `IO` implement `pos` `pos=` and `tell` in the base class to @asterite and simply provide a raising `IO#tell` by default.

For now, this is how `IO` works. It's not great, but it is how it is until I refactor.

File: src/crystal/system/unix/file_descriptor.cr (hunk 8 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: It wasn't documented before, I can add documentation if you want.

**Comment by RX14** [comment]
There have been loads of great ideas about refactoring `IO::FileDescriptor` and `IO` here, but I think that i'll work on a later PR which basically contains all this. I want to really focus on windows this christmas, instead of `IO` refactors.

**Comment by RX14** [comment]
Plus, a larger class means I have the chance to perform these same refactors on `Socket` and such. I really feel that these should be seperate PRs.

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: Personally i'd rather keep `pos`, `pos=` and `seek`. `tell` is obscure and weird naming to me.

File: src/crystal/system/unix/file_descriptor.cr (hunk 8 above)
Comment (ysbaddaden) [reply src/crystal/system/unix/file_descriptor.cr]: No, I meant: will the `#fcntl` method still appear in the documentation?

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (ysbaddaden) [reply src/crystal/system/unix/file_descriptor.cr]: Well, it's "seek and tell". They're well known associated terms, whatever the language. E.g. C, Haskell, Python, Ruby or Rust, all have both.

I don't like the `pos` naming, it's an abbreviation and inexplicit (inherited from Ruby). At least `position` would be a more explicit name but I still prefer `tell` (broadly used). Using a setter (`pos=`) to perform an action,  that is move the cursor in a file, sounds awful to me, in addition to be an exact duplicate for `seek`, too.

Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: I have to say "seek" is not an unknown term to me, although "tell" is. I agree with you on `pos` vs `position`, but not that setters shouldn't have performance side-effects.

File: src/crystal/system/unix/file_descriptor.cr (hunk 8 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: Hmm, not sure. Let me check.

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (straight-shoota) [reply src/crystal/system/unix/file_descriptor.cr]: As far as I can tell, the terms tell and seek come from working with files. `IO` is a general class, and to me it doesm't seem logical to transfer this meaning from files to any other stream. But I may be mistaken.
Another thought: `pos` and `pos=` is a concept easy to understand. One reads the current stream position, the other sets it (`position` and `position=` are a bit longer but also fine). With `seek` and `tell` it is more complicated, if your not familiar with these terms. You have to remember two names instead of one and can't simply asseses their meaning.

File: src/crystal/system/unix/file_descriptor.cr (hunk 8 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: They actually break the doc generator by saying they're included from the module, but the module is `:nodoc:` so that the links don't work.

Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: I'll go and do this method the same as the rest - with the `system_` prefix and `{% ifdef %}` the method out.

Comment (straight-shoota) [reply src/crystal/system/unix/file_descriptor.cr]: Yes, I've noticed this, too. Inclusion and extension of modules that are not documented should probably be hidden in the docs generator.

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (Papierkorb) [reply src/crystal/system/unix/file_descriptor.cr]: I'm also for dropping `#seek/#tell` and using `#position/#position=` (Not `#pos`).  The latter is really clear in intention.  Also, it lets you seek around in a file using normal code:

```cr
my_file.position += 5 # This is how you'd write it anywhere else too! Fantastic!
my_file.seek(my_file.tell + 5) # Compared to the one above: Meh.
my_file.seek(5, File::SEEK_CUR) # I can never remember which SEEK_* I need - this sucks.
```

File: src/crystal/system/unix/file_descriptor.cr (hunk 8 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: I pushed a fixup commit which resolves this. If this is OK i'll rebase to allow this to be merged.

File: src/crystal/system/unix/file_descriptor.cr (hunk 7 above)
Comment (RX14) [reply src/crystal/system/unix/file_descriptor.cr]: I think `seek` should stay, `io.position = io.size - 5` involves 2 syscalls whereas `io.seek(-5, IO::Seek::End)` requires only one. Similar for `io.position += 5`. That being said, I think that using `io.position` is a much nicer interface which is more than adequate for most people. Perhaps some caching for size and position would mean that we can obviate `seek` entirely.

Regardless of that, we should make a new issue containing all the IO refactor suggestions from this thread. Unless anyone else wants to do it now, i'll get it done once this PR is merged.

**Comment by RX14** [comment]
Ping? I think I've solved all the reviews, and all this should need is a tiny review of the fixup commit from @ysbaddaden.

**Comment by RX14** [comment]
Build failed with known failures related to travis's new build images: https://github.com/travis-ci/travis-ci/issues/8891.
